File: backend/internal/models/order.go
package models

import (
	"gomor-e-commerce/internal/repository"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type OrderItem struct {
	ProductID primitive.ObjectID `json:"productId" bson:"productId"`
	Name      string             `json:"name" bson:"name"`
	Quantity  int                `json:"quantity" bson:"quantity"`
	Image     string             `json:"image" bson:"image"`
	Price     float64            `json:"price" bson:"price"`
}

type ShippingAddress struct {
	Address    string `json:"address" bson:"address"`
	City       string `json:"city" bson:"city"`
	PostalCode string `json:"postalCode" bson:"postalCode"`
	Country    string `json:"country" bson:"country"`
}

type PaymentResult struct {
	ID           string `json:"id" bson:"id"`
	Status       string `json:"status" bson:"status"`
	UpdateTime   string `json:"update_time" bson:"update_time"`
	EmailAddress string `json:"email_address" bson:"email_address"`
}

type Order struct {
	repository.Auditable
	ID              *primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	UserID          string              `json:"userId" bson:"userId"`
	OrderItems      []OrderItem         `json:"orderItems" bson:"orderItems"`
	ShippingAddress ShippingAddress     `json:"shippingAddress" bson:"shippingAddress"`
	PaymentMethod   string              `json:"paymentMethod" bson:"paymentMethod"`
	PaymentResult   PaymentResult       `json:"paymentResult" bson:"paymentResult"`
	ItemsPrice      float64             `json:"itemsPrice" bson:"itemsPrice"`
	TaxPrice        float64             `json:"taxPrice" bson:"taxPrice"`
	ShippingPrice   float64             `json:"shippingPrice" bson:"shippingPrice"`
	TotalPrice      float64             `json:"totalPrice" bson:"totalPrice"`
	IsPaid          bool                `json:"isPaid" bson:"isPaid"`
	PaidAt          time.Time           `json:"paidAt" bson:"paidAt"`
	IsDelivered     bool                `json:"isDelivered" bson:"isDelivered"`
	DeliveredAt     time.Time           `json:"deliveredAt" bson:"deliveredAt"`
}

type OrderSalesTotal struct {
	ID    string  `json:"_id" bson:"_id"`
	Total float64 `json:"total" bson:"total"`
}

File: frontend/app/schemas/order.ts
import { z } from "zod";

export const orderItemSchema = z.object({
  name: z.string().min(1),
  quantity: z.number().int().min(1),
  image: z.string().min(1),
  price: z.number().min(0),
  productId: z.string(), // ObjectId as string
});

export const shippingAddressSchema = z.object({
  address: z.string().min(1),
  city: z.string().min(1),
  postalCode: z.string().min(1),
  country: z.string().min(1),
});

export const paymentResultSchema = z.object({
  id: z.string().optional(),
  status: z.string().optional(),
  update_time: z.string().optional(),
  email_address: z.string().email().optional(),
});

export const orderSchema = z.object({
  _id: z.string().optional(),
  userId: z.string(),
  orderItems: z.array(orderItemSchema),
  shippingAddress: shippingAddressSchema,
  paymentMethod: z.string().min(1),
  paymentResult: paymentResultSchema.optional(),
  itemsPrice: z.number().min(0).default(0),
  taxPrice: z.number().min(0).default(0),
  shippingPrice: z.number().min(0).default(0),
  totalPrice: z.number().min(0).default(0),
  isPaid: z.boolean().default(false),
  paidAt: z.string().datetime().optional(),
  isDelivered: z.boolean().default(false),
  deliveredAt: z.string().datetime().optional(),
  createdAt: z.iso.datetime().optional(),
  updatedAt: z.iso.datetime().optional(),
});

export type OrderItem = z.infer<typeof orderItemSchema>;
export type ShippingAddress = z.infer<typeof shippingAddressSchema>;
export type PaymentResult = z.infer<typeof paymentResultSchema>;
export type Order = z.infer<typeof orderSchema>;
